src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs

```diff
@@ -1025,6 +1041,38 @@ private void UpdateCompletedStreams()
             }
         }
 
+        // Compare to UpdateCompletedStreams, but only removes streams if over the max stream drain limit.
+        private void HandleStreamOverflow()
+        {
+            var now = SystemClock.UtcNowTicks;
+            var maxStreams = _serverSettings.MaxConcurrentStreams * 2;
+            // If we're tracking too many streams, discard the oldest.
+            while (_streams.Count > maxStreams && _completedStreams.TryDequeue(out var stream))
+            {
+                if (stream.DrainExpirationTicks == default)
+                {
+                    _serverActiveStreamCount--;
+                    stream.DrainExpirationTicks = now + Constants.RequestBodyDrainTimeout.Ticks;
+                }
+
+                if (stream == _currentHeadersStream)
+                {
+                    // This stream is currently draining trailers, hold onto it.
+                    _completedStreams.Enqueue(stream);
```

Comment:
Chatted offline: we probably don't need this re-enqueue if we don't enqueue the header stream initially.